src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -107,28 +107,49 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation)
         // Calculate cost due to {data point, target neighbors, impostors}
         // triplets.
         double eval = 0;
-        if ((iteration - 1 % range == 0) && (iteration - 1 != 0))
+        if (transformationOld.n_elem != 0)
         {
-          eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 distance(l, i);
+          eval = evalOld + arma::norm(transformation - transformationOld) *
+              (norm(targetNeighbors(j, i)) + norm(impostors(l, i)) +
+              2 * norm(i));
+          if (eval > -1)
+          {
+            // Calculate exact eval.
```

Comment:
I think this may be the reason for lack of speedup. Here each time` eval > -1` we are forced to do computation two times, one for eval bound and one for exact eval value.

Reply:
Yes, but your simulations in #1449 showed that a large number of the `eval` calculations could be skipped.  Have you tried finding out if we are skipping the `eval` calculation the expected number of times?

Reply:
I think I made some serious mistake while doing those simulations, I can't seem to reproduce those result. I even tried with previous 12  days versions of LMNN but all are giving same results, and are totally opposite to the simulations in #1449. According to current simulation, there are very few points on which eval < -1.

Reply:
I'm virtually certain your simulations were correct---so, under that assumption, what's different between what you did there and what's implemented here?

Reply:
I think I had done the same thing like I have initiated three count variables and after each eval calculation I have set three if else if conditions to increase the count based on eval value and just before return statement I am printing the values.

Reply:
Right, but you should be able to compute the number of evals as you did in #1449 by simply doing the same counting at the final check `if (eval <= -1) { bp = l; break; }`.

Reply:
that's what strange that the result is not the same. I even printed the eval values.

Reply:
I'd suggest starting the debugging from square one.  From my side I am quite confident that the original theory and your simulations in #1449 were correct (and this is supported by the observations in the JMLR paper).  So start from the beginning, make sure that you can get the original observations you were seeing in #1449 with unmodified code again, and then take a look at your implementation here and make sure that you still can.

Looking at the code my strong belief is that simply counting the number of `eval`s less than -2, less than -1, and greater than -1 in the code you have here should provide the same result.  But I believe that the bound implementation itself is wrong.  However, that is just my intuition, which is no substitute for digging into the problem and figuring out what is wrong.

Reply:
I couldn't get those results on iris and vc2 but with letters dataset, it is much better
```
(eval  > -1) --- (-2 < eval < -1) --- (eval < -2)
[DEBUG] Compiled with debugging symbols.
30240---28381---69168
[DEBUG] L-BFGS iteration 0; objective 124341, gradient norm 141545, 0.
13051---20816---77796
[DEBUG] L-BFGS iteration 1; objective 93106.6, gradient norm 44531.9, 0.2512.
5137---23661---75860
[DEBUG] L-BFGS iteration 2; objective 49849.1, gradient norm 19805.7, 0.464602.
30---22888---77110
[DEBUG] L-BFGS iteration 3; objective 22247.1, gradient norm 17154.7, 0.55371.
0---636---99364
[DEBUG] L-BFGS iteration 4; objective 9315.77, gradient norm 14277.2, 0.58126.
0---8---99992
[DEBUG] L-BFGS iteration 5; objective 5669.07, gradient norm 10134.7, 0.391455.
0---0---100000
```
I think I probably made some mistake earlier.

Reply:
It may be a little time consuming but I think it's very important to understand what is going on here.  So I think it is worthwhile to take the time to dig in further and see for sure if you made a mistake earlier or what.  Since I can't see the code you are using, I can't review it myself.  And with the code you just ran it looks a little suspicious to me since there are 0 evals greater than -1 by the fifth L-BFGS iteration.  That doesn't seem right to me.

During this whole process I would say it's really important to keep rigorous and detailed notes to avoid confusion like this, since if we want to eventually publish on it we need to be 100% sure that what we're writing is correct.

Reply:
Ideally, it should be like this: 
```
(eval  > -1)----(-2 < eval < -1)----(eval < -2)
1298----862----1590
944----656----2150
667----426----2657
921----647----2182
655----408----2687
898----636----2216
646----391----2713
866----633----2251
640----375----2735
```
But using these new bounds results are:
```
(eval  > -1)----(-2 < eval < -1)----(eval < -2)
0----0----3750
0----0----3750
0----0----3750
0----0----3750
0----0----3750
0----3750----0
0----3750----0
0----3750----0
0----3750----0
```
For cross - checking if I bypass our bounds by setting `if (transformationOld.n_elem != 0)` to false, then we are getting ideal results.